src/impl_/pyclass_init.rs

```diff
@@ -0,0 +1,93 @@
+//! Contains initialization utilities for `#[pyclass]`.
+use crate::ffi_ptr_ext::FfiPtrExt;
+use crate::internal::get_slot::TP_ALLOC;
+use crate::types::PyType;
+use crate::{ffi, Borrowed, PyErr, PyResult, Python};
+use crate::{ffi::PyTypeObject, sealed::Sealed, type_object::PyTypeInfo};
+use std::marker::PhantomData;
+
+/// Initializer for Python types.
+///
+/// This trait is intended to use internally for distinguishing `#[pyclass]` and
+/// Python native types.
+pub trait PyObjectInit<T>: Sized + Sealed {
+    /// # Safety
+    /// - `subtype` must be a valid pointer to a type object of T or a subclass.
+    unsafe fn into_new_object(
+        self,
+        py: Python<'_>,
+        subtype: *mut PyTypeObject,
+    ) -> PyResult<*mut ffi::PyObject>;
+
+    #[doc(hidden)]
+    fn can_be_subclassed(&self) -> bool;
+
+    private_decl! {}
```

Comment:
This is a different method of sealing, it can be removed now, together with `private_impl!()` below.

Reply:
It would actually be a nice change in this PR (or a follow up) to replace all uses of `private_decl!` / `private_impl!` with sealing, and remove that older method.

Reply:
Sure, that means some of the traits in pyclass need to be Sealed, which is what I plan to do next.